### PR TITLE
fix(sql): fix incorrect results caused by SQL optimiser pushing down some filters to nested SAMPLE BY aggregation

### LINKED ISSUE #2902: SQL optimiser moves timestamp filter inside SAMPLE BY screwing the results
### Describe the bug

_No response_

### To reproduce

When query runs as 

```SQL
with a as (
  select last(ts) as ts, ad, last(bl) as v
  from 'balance'
  where ts <= '2022-12-01T00:00:00.000000Z' and ad = 'AAA' and w = 'BBB' 
), b as (
  select ts, ad, bl as v
  from 'balance'
  where ts > '2022-12-01T00:00:00.000000Z' and ad = 'AAA' and w = 'BBB' 
), 
c as (select * FROM (select * from a union b) timestamp(ts) ),
d as (select ts, ad, last(v) as v from c sample by 1h fill(prev) align to calendar )
select ts as t, sum(v) as v, 'table_0' as group from d 
where ts >= '2022-12-01T00:00:00.000000Z'
```

The data returned is very different with `where ts >= '2022-12-01T00:00:00.000000Z'` and without `where ts >= '2022-12-01T00:00:00.000000Z'`
From logs query to execute is 

```

select-virtual t, v, 'table_0' group from (
        select-group-by [t, sum(v) v] t, sum(v) v from (
                select-choose [ts t, v] ts t, v from (
                        select-group-by [ts, last(v) v, ad] ts, ad, last(v) v from (
                                select-choose [ts, v, ad] ts, ad, v from (
                                        select [ts, v, ad] from (
                                                select-choose [ts, ad, v] ts, ad, v from (
                                                        select-group-by [last(ts) ts, ad, last(bl) v] last(ts) ts, ad, last(bl) v from (
                                                                select [ts, ad, bl, ad, workspace] from 
                                                                bl timestamp (ts) where ts <= '2022-12-01T00:00:00.000000Z' and ad = 'AAA' and workspace = 'BBB')) 
                                                a union 
                                                select-choose [ts, ad, v] ts, ad, v from (
                                                        select-choose [ts, ad, bl v] ts, ad, bl v from (
                                                                select [ts, ad, bl, ad, workspace] from 
                                                                bl timestamp (ts) where ts > '2022-12-01T00:00:00.000000Z' and ad = 'AAA' and workspace = 'BBB')
                                                        ) b
                                    ) _xQdbA1 where ts >= '2022-12-01T00:00:00.000000Z'
) timestamp (ts)) c sample by 1h fill(prev) align to calendar with offset '00:00') d) d) d
```

The filter `where ts >= '2022-12-01T00:00:00.000000Z'` is applied before SAMPLE BY and this changes the results since it impacts `fill(prev)` when there is no data for the first time interval 

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:6.6.1
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bziobrowski**
Fixes #2902

Disables predicate push down into sample by because it changes query result.
Timestamp produced by SAMPLE BY is not the same as source timestamp, it's rounded to some value .

When using : 
A. ALIGN TO FIRST OBSERVATION option - it's not possible to guess what the value is going to be. Predicates on non-timestamp columns can't be pushed down because they could alter first observation . 
B. ALIGN TO CALENDAR option -  it could be possible to translate timestamp in predicate to form compatible with SAMPLE BY (e.g. round down to day). 

This PR doens't attempt to keep predicate push down for B. because : 
- it wouldn't work for fill(prev) and fill(linear)
- it's unlikely to help with performance unless predicate can be translated into interval scan.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|6|6|100.00%|

**Review by puzpuzpuz — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|6|6|100.00%|

**Comment by ideoma** [comment]
Will it result to full table scans every time we have a `SAMPLE BY` in the query even though timestamp range is specified?

**Comment by bziobrowski** [comment]
No, it only affects predicates present in sample by's outer query . 
SampleByTest.testSampleByWithPredicate() shows that interval scan is used because ts filter is in Sample By's where clause .

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|6|6|100.00%|

**Comment by ideoma** [comment]
> No, it only affects predicates present in sample by's outer query . SampleByTest.testSampleByWithPredicate() shows that interval scan is used because ts filter is in Sample By's where clause .

Is it safe to push non timestamp filters down? I think it is. It can make a difference if a symbol filter is outside.
Also in some cases it safe to push ts filters too.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 9 / 14 (64.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|3|8|37.50%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|6|6|100.00%|

File: core/src/test/java/io/questdb/log/MockAlertTarget.java [hunk 1]
```diff
@@ -58,15 +64,21 @@ public void run() {
...
+                if (portNumber == 0) {
+                    portNumber = serverSkt.getLocalPort();
+                }
+
                 onTargetStart.run();
                 clientSkt = serverSkt.accept();
+
                 in = new BufferedReader(new InputStreamReader(clientSkt.getInputStream()));
                 out = new PrintWriter(clientSkt.getOutputStream(), true);
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/log/MockAlertTarget.java:76]: Reuse address combined with no linger created a 'storm' of packet retransmissions on windows that sometimes failed the test .

**Comment by bziobrowski** [comment]
> > No, it only affects predicates present in sample by's outer query . SampleByTest.testSampleByWithPredicate() shows that interval scan is used because ts filter is in Sample By's where clause .
> 
> Is it safe to push non timestamp filters down? I think it is. It can make a difference if a symbol filter is outside. Also in some cases it safe to push ts filters too.

@ideoma  
For : 
 - align to first observation 
  - pushing down any filter could alter first observation and thus whole result 
 - align to calendar
  - fill() other than none - any filter could change start ts or end ts or cut out values that would be used for later intervals (for prev and interpolate)      
  - fill(none) - predicate can only be pushed down if it aligns with sample by interval/time zone/offset . In most cases it'd need to be translated, not pushed down as is . 

Please have a look at tests I added to SampleByTest, try running it without and without change in SqlOptimiser and observe how outcome changes .

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 9 / 13 (69.23%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|3|7|42.86%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 9 / 13 (69.23%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|3|7|42.86%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|6|6|100.00%|

**Review by marregui — DISMISSED** [review]
LGTM

**Comment by ideoma** [comment]
> Please have a look at tests I added to SampleByTest, try running it without and without change in SqlOptimiser and observe how outcome changes .

I believe if there are no `FILL`s and data is aligned to calendar is safe to push down the filter of the columns which are grouped by. It can make the difference, run these on demo.questdb.io

```
select * from (
  select timestamp, symbol, side, first(price)
  from trades
  WHERE timestamp in '2022-12'
  SAMPLE by 1d ALIGN to CALENDAR
)
where symbol = 'BTC-USD' and side = 'sell'


  select timestamp, symbol, first(price)
  from trades
  WHERE timestamp in '2022-12' and symbol = 'BTC-USD' and side = 'sell'
  SAMPLE by 1d ALIGN to CALENDAR

```

The first query without external WHERE is 5x slower than the second query.

**Comment by bziobrowski** [comment]
I changed code to allow pushing non-ts predicates into sample by fill none align to calendar. 
Do you want me to also allow and rewrite predicates that use timestamp ?  e.g. 

```sql
select * from (
  select timestamp, symbol, side, first(price)
  from trades
  SAMPLE by 1d align to calendar
)
where timestamp in '2022-12' and symbol = 'BTC-USD' and side = 'sell'
->
select * from (
  select timestamp, symbol, side, first(price)
  from trades
  where timestamp_floor(timestamp, '1d') in '2022-12' 
  SAMPLE by 1d align to calendar
)
where symbol = 'BTC-USD' and side = 'sell'
```

Note:  this variant of sample by is basically the same as group by timestamp_floor(ts, unit), but group by is better at communicating that ts in output is not the same as ts in input .

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 40 / 46 (86.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|3|7|42.86%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|14|14|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 46 (89.13%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|3|7|42.86%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|24|25|96.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|14|14|100.00%|

**Comment by ideoma** [comment]
> Do you want me to also allow and rewrite predicates that use timestamp ? e.g.

It would be very good but there are restrictions when it's possible. Here for example it's not possible

```sql

select * from (
  select timestamp, symbol, side, first(price)
  from trades
  SAMPLE by 1d align to calendar
)
where timestamp > '2022-12-01T12' and symbol = 'BTC-USD' and side = 'sell'
```

It looks if you move `timestamp` filter inside you have to expand the range by sample by period

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 2]
```diff
@@ -734,6 +734,31 @@ private QueryModel bubbleUpOrderByAndLimitFromUnion(QueryModel model) throws Sql
...
+            return false;
+        }
+
+        CharSequence timestamp = findTimestamp(model);
+        if (timestamp == null) {
+            return true;
+        }
+
+        for (int i = 0, n = expressionColumns.size(); i < n; i++) {
+            if (Chars.equalsIgnoreCase(expressionColumns.get(i), timestamp)) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:754]: Will this catch something like 

```
select * from (
  select timestamp, symbol, side, first(price)
  from trades
  SAMPLE by 1d align to calendar
)
where symbol = 'BTC-USD' and side = 'sell' and dateadd('h', -1, timestamp) > '2022-12'
```

and prevent pushing down the timestamp filter?

This implementation is also is all or nothing which is not the best. In reality with the above query it's ok to push down `symbol` and `side` filters and leave timestamp filter externally.

Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/SqlOptimiser.java:754]: Please have a look at https://github.com/questdb/questdb/pull/2945/files#diff-5f98c23781bda3e3b7422770c5e1ef57d48e6ef6de4594abe849fea648d701e9 . 
This implementation is per where clause expression so it does push non-ts expressions if it makes sense .

**Comment by bziobrowski** [comment]
I don't think it makes sense to focus on this narrow optimization because there could be a large number of scenarios to consider (=,<,between,in,in with expression, ...) while at the same time we don't apply it when processing where caluse in general, e.g .
```sql
select * from trips where timestamp_floor('d', pickup_datetime) > '2021-01-01T00:11:00'
```
uses async filter instead of interval scan. 

A simple but safe rewrite would be to replace timestamp with 
`timestamp`
with 
`timestamp_floor( sample_by_unit, timestamp) `
when pushing down to sample by.  
Of course I'd need to extend timestamp_floor to support arbitrary units .

**Review by ideoma — APPROVED** [review]

**Pull request merged**